pruner/service.go

```diff
@@ -139,69 +141,88 @@ func (s *Service) run() {
 	}
 }
 
-func (s *Service) prune(
-	ctx context.Context,
-	lastPrunedHeader *header.ExtendedHeader,
-) *header.ExtendedHeader {
+func (s *Service) prune(ctx context.Context) {
+	// intentionally hold the lock for the whole pruning operation
+	s.checkpointMu.Lock()
+	defer s.checkpointMu.Unlock()
+
+	lastPrunedHeader, err := s.lastPruned(ctx)
+	if err != nil {
+		log.Errorw("getting last pruned header", "height", s.checkpoint.LastPrunedHeight, "err", err)
+		return
+	}
+
 	// prioritize retrying previously-failed headers
-	s.retryFailed(s.ctx)
+	s.retryFailed(ctx)
 
 	now := time.Now()
+	successful, failed := 0, 0
 	log.Debug("pruning round start")
 	defer func() {
-		log.Debugw("pruning round finished", "took", time.Since(now))
+		log.Infow("pruning round finished", "done", successful, "failed", failed, "took", time.Since(now))
+	}()
+
+	ctx, done := s.withWriteBatch(ctx)
+	defer func() {
+		if err := done(); err != nil {
+			log.Errorw("committing batch", "err", err)
+		}
 	}()
 
+	ctx, doneTxn := s.withReadTransaction(ctx)
```

Comment:
can you explain a bit about why we need this here too?

Reply:
Its an optional optimization

Reply:
If it's optional, does it need to be in this PR or can it come later?

Reply:
It could come later, obviously, but I don't see this being a big deal to copy and paste here. This has proven to be a substantial optimization that has minor maintenance overhead